content/blog/testthat-3-2-0/index.Rmd

```diff
@@ -0,0 +1,146 @@
+expect_in(x, c(TRUE, FALSE))
+```
+
+New `expect_no_error()`, `expect_no_warning()`, and `expect_no_message()` make it easier (and clearer) to expect that code runs with without errors, warnings, or messages.
+The default fails if there is any error/warning/message.
+In many cases, particularly when testing warnings and messages, you will want to be more specific about the condition you are hoping **not** to see, i.e. the condition that motivated you to write the test.
+Similar to `expect_error()` and friends, you can specify the `message` (a regular expression that the message of the condition must match) and/or the `class` (a class the condition must inherit from).
+This ensures that the message/warnings you don't want never recur, while allowing new messages/warnings to bubble up for you to deal with.
+
+## Snapshotting changes
+
+`expect_snapshot(error = TRUE)` has a new display of error messages that strives to be closer to what you see interactively.
+In particular, you'll no longer see the error class and you will now see the error call.
+
+If you have used `expect_snapshot(error = TRUE)` in your package, this means that you will need to re-run and approve your snapshots.
+We hope this is not too annoying and we believe it is worth it given the more accurate reflection of generated error messages.
+This will not affect checks on CRAN because, by default, snapshot tests are not run on CRAN.
+
+## State inspector
+
+In times gone by it was very easy to accidentally change the state of the world in a test:
+
+```{r}
+#| eval: false
+test_that("side-by-side diffs work", {
+  options(width = 20)
+  expect_snapshot(
+    waldo::compare(c("X", letters), c(letters, "X"))
+  )
+})
+```
+
+When you look at a single test it's easy to spot the problem, and switch to a more appropriate way of temporarily changing the options, like `withr::local_options()`.
+But sometimes this mistake crept in a long time ago and is now hiding amongst hundreds or thousands of tests.
+
+In earlier versions of testthat, finding tests that accidentally changed the world was painful: the only way was to painstakingly review each test.
+Now you can use `set_state_inspector()` to register a function that's called before and after every test.
+The function returns different values, testthat will let you know.
+
+So, for example, to detect if any of your tests have modified options you could put the following in `tests/testhat/setup.R` or an existing helper file:
+
+```{r}
+set_state_inspector(function() {
+  list(options = options())
+})
+```
+
+Or maybe you've an `R CMD check` warning that you've forgotten to close a connection:
+
+```{r}
+set_state_inspector(function() {
+  list(connections = nrow(showConnections()))
+})
+```
+
+And you can of course combine multiple checks just by returning a more complicated list.
+
+## UI improvements
+
+testthat 3.2.0 includes a bunch of minor user interface improvements that should make day-to-day use of testthat more enjoyable.
```

Comment:
I faintly recall you saying that you believe parallel testing is now much more robust, do you want to mention it?

Reply:
Indeed there weren't any changes in parallel testing from the first implementation AFAIR, except that it interacts better with snapshots now, and testthat deletes unneeded snapshots files.

Reply:
I think the main parallel-related fix happened outside of testthat? IIRC there was an issue with processes sometimes getting stuck / left behind if you terminated `devtools::test()` while it was still trying to run tests.

Reply:
Mentioned both points.